tempesta_fw/http_msg.c

```diff
@@ -819,3 +848,89 @@ tfw_http_msg_alloc(int type)
 
 	return hm;
 }
+
+static inline void
+__mark_hbh_header(TfwStr * hdr)
+{
+	if (!(TFW_STR_EMPTY(hdr)))
+		hdr->flags |= TFW_STR_HBH_HDR;
+}
+
+/**
+ * Locate and mark all Hop-by-Hop headers by setting TFW_STR_HBH_HDR
+ * flag to corresponding TfwStr's in header table.
+ */
+void tfw_http_msg_mark_hbh_hdrs(TfwHttpMsg *hm)
+{
+	TfwStr conn_h, hbh_hdr;
+	TfwStr *c, *end;
+	unsigned int id;
+
+	/*
+	 *  Server header isn't defined as hop-by-hop by the RFC, but we
+	 * don't show protected server to world.
+	 */
+	__mark_hbh_header(&hm->h_tbl->tbl[TFW_HTTP_HDR_SERVER]);
+	__mark_hbh_header(&hm->h_tbl->tbl[TFW_HTTP_HDR_CONNECTION]);
+	__mark_hbh_header(&hm->h_tbl->tbl[TFW_HTTP_HDR_KEEP_ALIVE]);
+
+	if (!(hm->flags & TFW_HTTP_CONN_EXTRA))
+		return;
+
+	tfw_http_msg_srvhdr_val(&hm->h_tbl->tbl[TFW_HTTP_HDR_CONNECTION],
+				TFW_HTTP_HDR_CONNECTION, &conn_h);
+
+	TFW_STR_FOR_EACH_CHUNK(c, &conn_h, end) {
+		if (c->flags & TFW_STR_VALUE) {
+			TFW_STR_INIT(&hbh_hdr);
+			hbh_hdr.ptr = c;
+			TFW_STR_CHUNKN_ADD(&hbh_hdr, 1);
+			while ((c->flags & TFW_STR_VALUE)
+			       && (c != end)) {
+				hbh_hdr.len += c->len;
+				++c;
+			}
+
+			id = __hdr_lookup_hdr_name(hm, &hbh_hdr);
+			if (id < hm->h_tbl->off) {
+				hm->h_tbl->tbl[id].flags |= TFW_STR_HBH_HDR;
+			}
+		}
+	}
+	/*
+	 * Unset TFW_HTTP_CONN_EXTRA flag:
+	 * @tfw_http_msg_mark_hbh_hdrs may be called more than once for same
+	 * message. Since marks stay in http message until it would be
+	 * destroyed, we can reuse marks later without heavy processing.
+	 */
+	hm->flags &= ~((unsigned int)TFW_HTTP_CONN_EXTRA);
```

Comment:
This serves as an indicator that allows to NOT perform heavy marking again when it had been done already. However, why not avoid repeated calls to mark the headers altogether? That can be done by setting a flag once the marking is done.